jubatus/core/classifier/nearest_neighbor_classifier.cpp

```diff
@@ -256,6 +260,29 @@ void nearest_neighbor_classifier::unlearn_id(const std::string& id) {
   nearest_neighbor_engine_->get_table()->delete_row(id);
 }
 
+void nearest_neighbor_classifier::unlearn_label(const std::string& label) {
+  util::concurrent::scoped_lock lk(label_mutex_);
+  labels_[label] -= 1;
+  if (labels_[label] <= 0) {
+    labels_.erase(label);
+  }
+}
+
+void nearest_neighbor_classifier::regenerate_labels() {
+  shared_ptr<const storage::column_table> table =
+      nearest_neighbor_engine_->get_const_table();
+
+  labels_t labels;
+  for (size_t i = 0, n = table->size(); i < n; ++i) {
+    std::string id = table->get_key(i);
+    std::string label = get_label_from_id(id);
+    labels[label] += 1;
+  }
+
+  util::concurrent::scoped_lock lk(label_mutex_);
+  labels_.swap(labels);
```

Comment:
Labels added via `set_label` API will be lost here.  Copying keyset from `labels_` seems better.